app/src/test/kotlin/compiler/intermediate_form/LinearizationTest.kt

```diff
@@ -0,0 +1,437 @@
+            listOf(node1, node2, node3),
+            node1,
+            referenceHashMapOf(),
+            referenceHashMapOf(node1 to node2, node2 to node2),
+            referenceHashMapOf(node1 to node3, node2 to node1)
+        )
+
+        val result = Linearization.linearize(cfg, covering)
+
+        assertLinearizationMatches(
+            listOf(
+                NextLabel,
+                Conditional(node1, false, 2),
+                NextLabel,
+                Conditional(node2, true, 1),
+                Jump(0),
+                NextLabel,
+                Uncoditional(node3)
+            ),
+            result
+        )
+    }
+
+    @Test
+    fun `end if true`() {
+        val node1 = newNode()
+        val node2 = newNode()
+
+        val cfg = ControlFlowGraph(
+            listOf(node1, node2),
+            node1,
+            referenceHashMapOf(),
+            referenceHashMapOf(),
+            referenceHashMapOf(node1 to node2)
+        )
+
+        val result = Linearization.linearize(cfg, covering)
+
+        assertLinearizationMatches(
+            listOf(
+                Conditional(node1, true, 0),
+                Uncoditional(node2),
+                NextLabel
+            ),
+            result
+        )
+    }
+
+    @Test
+    fun `end if false`() {
+        val node1 = newNode()
+        val node2 = newNode()
+
+        val cfg = ControlFlowGraph(
+            listOf(node1, node2),
+            node1,
+            referenceHashMapOf(),
+            referenceHashMapOf(node1 to node2),
+            referenceHashMapOf()
+        )
+
+        val result = Linearization.linearize(cfg, covering)
+
+        assertLinearizationMatches(
+            listOf(
+                Conditional(node1, false, 0),
+                Uncoditional(node2),
+                NextLabel
+            ),
+            result
+        )
+    }
+
+    @Test
+    fun `end after empty loop checking if true`() {
+        val node = newNode()
+
+        val cfg = ControlFlowGraph(
+            listOf(node),
+            node,
+            referenceHashMapOf(),
+            referenceHashMapOf(node to node),
+            referenceHashMapOf()
+        )
+
+        val result = Linearization.linearize(cfg, covering)
+
+        assertLinearizationMatches(
+            listOf(
+                NextLabel,
+                Conditional(node, true, 0)
+            ),
+            result
+        )
+    }
+
+    @Test
+    fun `end after empty loop checking if false`() {
+        val node = newNode()
+
+        val cfg = ControlFlowGraph(
+            listOf(node),
+            node,
+            referenceHashMapOf(),
+            referenceHashMapOf(),
+            referenceHashMapOf(node to node)
+        )
+
+        val result = Linearization.linearize(cfg, covering)
+
+        assertLinearizationMatches(
+            listOf(
+                NextLabel,
+                Conditional(node, false, 0)
+            ),
+            result
+        )
+    }
+}
```

Comment:
Tests are pretty exhaustive but maybe we could add something with nested loops or conditions?

Reply:
Added two more tests. (Note that the tests had already covered all lines in `Linearization`.)